framework/fp_vectors/Floats.h

```diff
@@ -8,34 +8,230 @@
+#define FP16_HAS_QUIET_NAN      1
+#define FP16_MANT_DIG           11
+#define FP16_MAX_10_EXP         4
+#define FP16_MAX                6.55040000000000000000000000000000000e+4
+#define FP16_MAX_EXP            16
+#define FP16_MIN_10_EXP         (-4)
+#define FP16_MIN                6.10351562500000000000000000000000000e-5
+#define FP16_MIN_EXP            (-13)
+#define FP16_NORM_MAX           6.55040000000000000000000000000000000e+4
+
+#define BFLT16_DECIMAL_DIG      3
+#define BFLT16_DENORM_MIN       (0x1p-133)
+#define BFLT16_DIG              2
+#define BFLT16_EPSILON          (FLT_EPSILON * 65536)
+#define BFLT16_HAS_DENORM       1
+#define BFLT16_HAS_INFINITY     1
+#define BFLT16_HAS_QUIET_NAN    1
+#define BFLT16_MANT_DIG         (FLT_MANT_DIG - 16)
+#define BFLT16_MAX_10_EXP       FLT_MAX_10_EXP
+#define BFLT16_MAX_EXP          FLT_MAX_EXP
+#define BFLT16_MAX              (0x1.fep+127f)
+#define BFLT16_MIN_10_EXP       FLT_MIN_10_EXP
+#define BFLT16_MIN_EXP          FLT_MIN_EXP
+#define BFLT16_MIN              (0x1p-126f)
+#define BFLT16_NORM_MAX         BFLT16_MAX
+
 #ifdef __cplusplus
 extern "C" {
 #endif
 
+struct Float16
+{
+    union {
+#if defined(SANDSTONE_HAS_FLOAT16_TYPE) || \
+    defined(SANDSTONE_HAS__FP16_TYPE)
+
+        fp16_t as_float;
+
+#endif
+
+        union {
+            uint16_t as_hex;
+            uint16_t payload;
+        };
+
+        struct __attribute__((packed)) {
+            uint16_t mantissa : 10;
+            uint16_t exponent : 5;
+            uint16_t sign : 1;
+        };
+    };
+
+#ifdef __cplusplus
+    Float16() = default;
+    inline Float16(float f);
+
+    static constexpr int digits = FP16_MANT_DIG;
+    static constexpr int digits10 = FP16_DIG;
+    static constexpr int max_digits10 = 6;  // log2(digits)
+    static constexpr int min_exponent = FP16_MIN_EXP;
+    static constexpr int min_exponent10 = FP16_MIN_10_EXP;
+    static constexpr int max_exponent = FP16_MAX_EXP;
+    static constexpr int max_exponent10 = FP16_MAX_10_EXP;
+
+    static constexpr bool radix = 2;
+    static constexpr bool is_signed = true;
+    static constexpr bool is_integer = false;
+    static constexpr bool is_exact = false;
+    static constexpr bool has_infinity = FP16_HAS_INFINITY;
+    static constexpr bool has_quiet_NaN = FP16_HAS_QUIET_NAN;
+    static constexpr bool has_signaling_NaN = has_quiet_NaN;
+    static constexpr std::float_denorm_style has_denorm = std::denorm_present;
+    static constexpr bool has_denorm_loss = false;
+    static constexpr bool is_iec559 = true;
+    static constexpr bool is_bounded = true;
+    static constexpr bool is_modulo = false;
+    static constexpr bool traps = false;
+    static constexpr bool tinyness_before = false;
+    static constexpr std::float_round_style round_style =
+            std::round_toward_zero;   // unlike std::numeric_limits<float>::round_style
+
+    static constexpr Float16 min()              { return Float16(Holder{0x0400}); }
```

Comment:
calling something min() that isn't min(a,b) is a little... surprise causing

Reply:
I don't think we can make ISO change their minds ;p
https://en.cppreference.com/w/cpp/types/numeric_limits/min

It is made that way so we have the same API as std::numeric_limits.